Leveraging React Context Across Render Boundaries in Next.js - Lessons from Building a Locale Switcher

timgruenewald opened the pull request:
In Next.js applications, layouts often need access to data that only becomes available after a nested page has rendered - for example, localized slugs from a headless CMS that are required to build SEO-friendly, localized URLs. This article explores how to bridge that gap with React Context, why global stores and "server contexts" are usually not a good idea, and how a context-based pattern can help manage your state safely across render client/server boundaries.

timgruenewald committed:
feat: add article on leveraging React Context for locale switching in Next.js applications

- Introduced a comprehensive guide on building a locale switcher using React Context in Next.js.
- Discussed challenges related to accessing localized slugs and managing state across render boundaries.
- Implemented a Context Hydrator pattern to effectively bridge server-rendered data with client-side components.
- Provided code examples for context setup, hydrator component, and locale switcher functionality.

timgruenewald committed:
fix: improve readability of introduction section in locale switcher article

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+  - Context
+shortDescription: >-
+  Implement smooth fade page transitions in React apps using react-router v6 and
+  framer motion. Learn to handle component lifecycle for seamless animations.
```
**Wrong `shortDescription`** — this appears to be copied from a completely unrelated article (*react-router-6-page-transitions*):

> Implement smooth fade page transitions in React apps using react-router v6 and framer motion. Learn to handle component lifecycle for seamless animations.

Please replace with a short description that accurately summarises this article — e.g. something like:

> Learn how to bridge server-rendered CMS data into a client-side locale switcher in Next.js using a React Context hydrator pattern, without global stores or unstable server context APIs.

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+# Introduction
+
+On the project I am currently working on, we are building a Next.js powered web application for a health care provider.
+Some content we show in the web app such as product data or user specific information such as active subscriptions to services is curated via a admin UI, though the main part of the content comes from pages and blog articles as well as images that are fully driven by [Dato, a headless CMS.](https://www.datocms.com/).
```
"a admin UI" → **"an admin UI"** (the article "a" must become "an" before a vowel sound).

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+
+On the project I am currently working on, we are building a Next.js powered web application for a health care provider.
+Some content we show in the web app such as product data or user specific information such as active subscriptions to services is curated via a admin UI, though the main part of the content comes from pages and blog articles as well as images that are fully driven by [Dato, a headless CMS.](https://www.datocms.com/).
+Consequentially, navigation elements and footer content are curated in the CMS as well, with links representing references to other CMS-driven data records.
```
"Consequentially" → **"Consequently"**. "Consequentially" means "of great consequence / significance"; the intended meaning here is "as a result / therefore", which is "consequently".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+Some content we show in the web app such as product data or user specific information such as active subscriptions to services is curated via a admin UI, though the main part of the content comes from pages and blog articles as well as images that are fully driven by [Dato, a headless CMS.](https://www.datocms.com/).
+Consequentially, navigation elements and footer content are curated in the CMS as well, with links representing references to other CMS-driven data records.
+
+In order to pave the way for a fully localized user experience, we had to implement a locale resp. language switcher component located in the header of the web application.
```
"locale resp. language switcher" — `resp.` is a German abbreviation (*respektive*), not standard English. Replace with **"or"**: "locale or language switcher component".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+In order to pave the way for a fully localized user experience, we had to implement a locale resp. language switcher component located in the header of the web application.
+This dropdown component should always be able to serve the correct, localized URL to the currently viewed page in the other languages the content is available in.
+Now, this would be trivial if the task at hand would be to just switch the locale with the rest of the URL staying the same, e.g. linking from a German article located under `/de/blog/gesund-essen` to the English version of it `/en/blog/gesund-essen`.
+But as [Google documents on "Google Search Central"](developers.google.com/search/docs/crawling-indexing/url-structure#use-your-audiences-language), the best practice when it comes to localized URLs is to use human-readable, descriptive URLs in your audience's language (e.g. in our example `/en/healthy-eating`).
```
The link target is missing the `https://` scheme: `developers.google.com/search/...` — this will not resolve as a clickable hyperlink. Change to:

```
[Google documents on "Google Search Central"](https://developers.google.com/search/docs/crawling-indexing/url-structure#use-your-audiences-language)
```

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+Now, this would be trivial if the task at hand would be to just switch the locale with the rest of the URL staying the same, e.g. linking from a German article located under `/de/blog/gesund-essen` to the English version of it `/en/blog/gesund-essen`.
+But as [Google documents on "Google Search Central"](developers.google.com/search/docs/crawling-indexing/url-structure#use-your-audiences-language), the best practice when it comes to localized URLs is to use human-readable, descriptive URLs in your audience's language (e.g. in our example `/en/healthy-eating`).
+
+In the following article I will talk about the challenges of building such a locale switcher component if it requires CMS-driven data might only be only available after the component itself has already finished rendering, a problem touching a lot of different aspects and pitfalls of building projects with Next.js, and how our solution to this both leverages the inner workings of React context and ...
```
Duplicate word: "data might only be **only** available" — remove one "only".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+    <ul>
+      {articles.map((article) => (
+        <li>
+          <Link href={`/${locale}/blog/${slug}`}>{article.title}</Link>
```
`slug` is not in scope here — this code will not compile. `slug` is not a property of the `params` for `ArticleListPage` (which only has `locale`). You likely want `article.slug` (or whatever field the article object exposes for its slug):

```tsx
<Link href={`/${locale}/blog/${article.slug}`}>{article.title}</Link>
```

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+A layout is shared UI that wraps multiple pages. The key benefit of layouts is that they preserve state, remain interactive, and don’t re-render when a user navigates between pages that share the layout.
+
+The handling of the locale and internationalization, including routing, correctly switching localized strings stored in resource files as well as providing us with several useful hooks is handled by [next-intl](https://next-intl.dev/).
+You don't have to worry about the details, just note here that we are wrapping our pages in a `NextIntlClientProvider`, a context provider that comes with the plugin and enables the beforementioned hooks to access the current locale, regardless of where they are located in the render tree, as long as they are a child component of this provider and a client-side component.
```
"beforementioned" is not standard English. Use **"aforementioned"**: "enables the aforementioned hooks to access the current locale".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+
+Regardless of what our locale switcher component looks like, it will need at least two key pieces of information to correctly resolve the URLs of our article page for other languages if we want to give users a fully localized experience:
+
+1. The current pathname resp. article page we are on, e.g. `/blog/healthy-eating`
```
"the current pathname resp. article page" — `resp.` is a German abbreviation, not English. Replace with **"or"**: "the current pathname or article page we are on".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
````diff
@@ -0,0 +1,531 @@
+   ```
+
+   The information about the article's slugs in other languages is contained here in the `article` object.
+   But how do we get this information up into the layout and into the layout?
````
Repeated phrase: "But how do we get this information up **into the layout and into the layout**?" — this appears to be a copy-paste error. Did you mean "up into the layout and into the locale switcher"?

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
````diff
@@ -0,0 +1,531 @@
+$localizedParams.set(initialValuesFromServer);
+```
+
+An obvious solution to this would be to make the locale switcher a server component as well, but even that that wouldn't solve the problem, because:
````
Duplicate word: "but even **that that** wouldn't solve the problem" — remove one "that".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+
+React Server Components are parallelized and streamed, so a page cannot send data “upward” after the layout tree has already been serialized.
+
+Even withing a single request / render cycle, Next.js doesn't necessarily guarantee an order of execution when it comes to rendering layouts and pages.
```
Typo: "Even **withing** a single request" → **"within"**.

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+4. **Nanostores stores are not scoped to a single render cycle**
+
+This was a small caveat in the face of the other, more fundamental problems we ran into with trying to solve this with server side stores, but it was still a core problem: Nanostores stores are not isolated to a single render cycle by default.
+They are global in the sense of that they are tied to the life cycle of the node process running the Next.js server, so you definitely run the risk of leaking data between different render cycles and requests, but also between different users.
```
Awkward phrasing: "global in the **sense of that** they are tied" — should be "in the **sense that** they are tied".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+
+Now you may ask yourself:
+
+# What about React Server Context
```
This is a question and should end with a question mark: `# What about React Server Context?`

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+
+# React's cache()
+
+Facing the problem with React server context, React introduced react.cache() as a safer, more predictable primitive for (some of) its use cases.
```
"React introduced **react.cache()**" — the namespace should be capitalised: **`React.cache()`**. Same applies on the bullet point below. (The actual import is `import { cache } from 'react'`, so in prose the correct reference is `React.cache()` or simply `cache()`.)

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+
+# Moving away from the idea of "server context"
+
+To summarize the last sections about different approaches, attempts and their pitfalls: Server context is generally a bad idea, at least in the context of React and Next.js for several reasons:
```
British English: "To **summarize**" → "To **summarise**".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+  It couples components to hidden global state, reduces reusability outside that context, and makes debugging harder.
+
+- **Essentially introduces side effects:**
+  Server Context turns server rendering into a stateful environment, which breaks the mental model of React Server Components and leads to unpredictable behavior.
```
British English: "unpredictable **behavior**" → "unpredictable **behaviour**".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+- **Essentially introduces side effects:**
+  Server Context turns server rendering into a stateful environment, which breaks the mental model of React Server Components and leads to unpredictable behavior.
+
+# The Solution : A Context Hydrator Pattern
```
Extra space before the colon in the heading: `# The Solution : A Context Hydrator Pattern` → `# The Solution: A Context Hydrator Pattern`.

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+
+Instead, the page simply hands the data to the browser, and the browser hands it to the layout subtree through context.
+
+At a first glance, this solution might seem counter-intuitive and I'd argue it even is, because most tutorials and even the React documentation focus mostly on how to use Context as a way to move state upwards the component tree and avoid prop-drilling.
```
"**At a first glance**" — the "a" is superfluous. Use **"At first glance"**.

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+import { createContext, useContext, useEffect, useState } from "react";
+import { SiteLocale } from "@/app/types/SiteLocale";
+
+type SlugMap = Record<SiteLocale, string>>;
```
Two bugs on this line:

1. **Syntax error** — the double `>>` is invalid TypeScript syntax and will cause a compile error.
2. **Wrong value type** — looking at how `setSlugsState` merges values (the `setSlugs` function merges an object per locale), the value type must be `Record<string, string>`, not `string`.

```ts
// ✅ Correct
type SlugMap = Record<SiteLocale, Record<string, string>>;
```

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+
+      setSlugs(locale as SiteLocale, { [paramKey]: value });
+    });
+  }, [values, paramKey, getSlugs, setSlugs]);
```
`getSlugs` and `setSlugs` are plain functions re-created on every render of `LocalizedSlugsProvider`, so listing them as `useEffect` dependencies will cause this effect to fire on every render, not just when `values` or `paramKey` change.

The fix is to wrap `getSlugs` and `setSlugs` in `useCallback` inside the provider, so their references are stable:

```ts
const setSlugs = useCallback((locale: SiteLocale, newSlugs: Record<string, string>) => {
  setSlugsState((prev) => ({ ...prev, [locale]: { ...(prev[locale] || {}), ...newSlugs } }));
}, []);

const getSlugs = useCallback((locale: SiteLocale) => slugs[locale], [slugs]);
```

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
````diff
@@ -0,0 +1,531 @@
+```
+
+The `LocalizedSlugsProvider` creates a simple React Context that stores all localized slugs, grouped by locale.
+It exposes setter functions so the client can update this data later.
````
"It exposes **setter functions** so the client can update this data later." — the context also exposes a getter (`getSlugs`). Should read "**getter and setter functions**".

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+// app/[locale]/blog/[slug]/page.tsx
+import { LocalizedSlugsHydrator } from "../../components/LocalizedSlugsHydrator";
+
+export default async function ProductPage({ params }) {
```
`params` is missing its TypeScript type annotation. Every other page component in this article is correctly typed — this example should be consistent:

```tsx
export default async function ProductPage({ params }: { params: Promise<{ locale: string; slug: string }> }) {
```

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+const LOCALES: SiteLocale[] = ["de", "fr", "it"];
+
+export function LocaleSwitcher() {
+  const pathname = usePathname();
```
`usePathname` and `Link` are used in this snippet but neither is imported. For readers new to Next.js these are particularly non-obvious. Please add:

```tsx
import { usePathname } from "next/navigation";
import Link from "next/link";
```

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
```diff
@@ -0,0 +1,531 @@
+
+import { useLocalizedSlugs } from "@/context/LocalizedSlugsContext";
+
+const LOCALES: SiteLocale[] = ["de", "fr", "it"];
```
The `LOCALES` array hardcodes `["de", "fr", "it"]` (three locales including Italian), but the article's running example throughout only ever mentions German and French. Either align the array with the prose example (`["de", "fr"]`) or introduce Italian somewhere in the narrative earlier.

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
**British English spelling — "localize" family** — the article consistently uses American English spelling throughout. As the blog uses British English, all of these need updating:

| American (current) | British (required) |
|---|---|
| localized | localised |
| localizing | localising |
| localization | localisation |
| localize | localise |

This affects around 10 occurrences across prose, the frontmatter `description`, and headings. (Occurrences inside code blocks/backtick literals should remain as-is, since that reflects the actual variable/function names.)

kunman93 commented on src/articles/react-context-nextjs-locale-switcher/index.md:
**Heading hierarchy** — the sections "What about React Server Context?", "React's cache()", and "Moving away from the idea of 'server context'" are all logically sub-topics of the failed-approaches narrative started by "First Attempt: Global (Server-Side) Stores", yet all four share the same top-level `#` heading. This flattens the article structure and makes it harder for readers to follow the flow.

Consider demoting those three sections to `##` and nesting them under a broader top-level section such as `# Why Common Approaches Fall Short`, so the structure reads:

```
# The Challenge
# Why Common Approaches Fall Short
  ## First Attempt: Global (Server-Side) Stores
  ## What about React Server Context?
  ## React's cache()
  ## Moving away from the idea of "server context"
# The Solution: A Context Hydrator Pattern
# Key Lessons
# Conclusion
```

kunman93 reviewed (changes requested):
Good article overall — the problem is clearly motivated and the solution pattern is well explained. I've left 22 inline comments covering the following categories:

**Must fix:**
- Wrong `shortDescription` in frontmatter (copy-pasted from a different article)
- Broken hyperlink (missing `https://` scheme on the Google Search Central URL)
- Two undefined-variable / compile errors in code examples
- TypeScript syntax error in `SlugMap` type (`>>`)
- Missing `usePathname` and `Link` imports in the `LocaleSwitcher` snippet
- Unstable `useEffect` dependency references (`getSlugs`, `setSlugs`)

**Language (British English):**
- American → British spelling throughout (`localized` → `localised`, `behavior` → `behaviour`, `summarize` → `summarise`)
- `Consequentially` → `Consequently`
- `resp.` → `or` (German abbreviation, not English)
- `beforementioned` → `aforementioned`
- `at a first glance` → `at first glance`

**Typos / duplicate words:**
- "only be only" — duplicate "only"
- "even that that" — duplicate "that"
- "withing" → "within"
- "into the layout and into the layout" — repeated phrase
- "in the sense of that" → "in the sense that"

**Structure:**
- Section heading missing `?` (`# What about React Server Context`)
- Space before colon in heading (`# The Solution : …`)
- `react.cache()` → `React.cache()`
- Flat heading hierarchy across the failed-approaches sections
- `LOCALES` constant inconsistent with running prose examples

timgruenewald committed:
chore(article): polish locale-switcher post for review feedback

- fix frontmatter short description and broken Google Search Central link
- correct code-snippet issues (undefined variable, SlugMap type, missing imports)
- improve snippet reliability by stabilizing hydrator effect dependencies
- clean up typos, duplicate words, and awkward phrasing
- normalize heading punctuation/style and improve section hierarchy
- align LOCALES example with prose and keep prose in American English

timgruenewald committed:
fix: update diagram format to use Mermaid syntax for better clarity in locale switcher article

timgruenewald committed:
fix: correct typos and improve clarity in locale switcher article

timgruenewald committed:
fix: enhance clarity in locale switcher article by restructuring sentences and improving section key assignment

timgruenewald committed:
Merge branch 'main' into react-context-nextjs-locale-switcher

timgruenewald committed:
fix: update Mermaid diagram syntax for improved clarity in locale switcher article

yasmineantille reviewed (approved):
Approved since Tim implemented Manu's feedback.
